File: example/src/oklog/java/com/github/simonpercic/oklogexample/data/FlavorDataModule.java
package com.github.simonpercic.oklogexample.data;

import com.github.simonpercic.oklog.OkLogInterceptor;
import com.github.simonpercic.oklogexample.Constants;
import com.github.simonpercic.oklogexample.data.api.ApiService;
import com.github.simonpercic.oklogexample.data.api.RestApi;
import com.github.simonpercic.oklogexample.data.api.RestApiService;
import com.squareup.okhttp.Interceptor;
import com.squareup.okhttp.OkHttpClient;

import java.util.List;
import java.util.Set;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import dagger.multibindings.IntoSet;
import retrofit.Endpoints;
import retrofit.RestAdapter;
import retrofit.client.OkClient;

/**
 * @author Simon Percic <a href="https://github.com/simonpercic">https://github.com/simonpercic</a>
 */
@Module
public class FlavorDataModule {

    @Provides @Singleton ApiService provideApiService(RestApiService restApiService) {
        return restApiService;
    }

    @Provides @Singleton RestApi provideRestApi(OkHttpClient okHttpClient) {
        return new RestAdapter.Builder()
                .setClient(new OkClient(okHttpClient))
                .setEndpoint(Endpoints.newFixedEndpoint(Constants.ENDPOINT))
                .setLogLevel(RestAdapter.LogLevel.FULL)
                .build()
                .create(RestApi.class);
    }

    @Provides @IntoSet @Singleton Interceptor provideOkLogInterceptor() {
        return OkLogInterceptor.builder().build();
    }

    @Provides @Singleton OkHttpClient provideOkHttpClient(Set<Interceptor> interceptors) {
        OkHttpClient client = new OkHttpClient();

        List<Interceptor> clientInterceptors = client.interceptors();

        if (interceptors != null && interceptors.size() > 0) {
            for (Interceptor interceptor : interceptors) {
                clientInterceptors.add(interceptor);
            }
        }

        return client;
    }
}

File: example/src/oklog/kotlin/com/github/simonpercic/oklogexample/data/FlavorDataModule.kt
package com.github.simonpercic.oklogexample.data

import com.github.simonpercic.oklog.OkLogInterceptor
import com.github.simonpercic.oklogexample.Constants
import com.github.simonpercic.oklogexample.data.api.ApiService
import com.github.simonpercic.oklogexample.data.api.RestApi
import com.github.simonpercic.oklogexample.data.api.RestApiService
import com.squareup.okhttp.Interceptor
import com.squareup.okhttp.OkHttpClient
import dagger.Module
import dagger.Provides
import dagger.multibindings.IntoSet
import retrofit.Endpoints
import retrofit.RestAdapter
import retrofit.client.OkClient
import javax.inject.Singleton

/**
 * @author Simon Percic [https://github.com/simonpercic](https://github.com/simonpercic)
 */
@Module
class FlavorDataModule {

    @Provides @Singleton internal fun provideApiService(restApiService: RestApiService): ApiService {
        return restApiService
    }

    @Provides @Singleton internal fun provideRestApi(okHttpClient: OkHttpClient): RestApi {
        return RestAdapter.Builder()
                .setClient(OkClient(okHttpClient))
                .setEndpoint(Endpoints.newFixedEndpoint(Constants.ENDPOINT))
                .setLogLevel(RestAdapter.LogLevel.FULL)
                .build()
                .create(RestApi::class.java)
    }

    @Provides @IntoSet @Singleton internal fun provideOkLogInterceptor(): Interceptor {
        return OkLogInterceptor.builder().build()
    }

    @Provides @Singleton internal fun provideOkHttpClient(interceptors: Set<@JvmSuppressWildcards Interceptor>?): OkHttpClient {
        val client = OkHttpClient()

        val clientInterceptors = client.interceptors()

        if (interceptors != null && interceptors.isNotEmpty()) {
            for (interceptor in interceptors) {
                clientInterceptors.add(interceptor)
            }
        }

        return client
    }
}
